src/languages/r.js

```diff
@@ -2,76 +2,148 @@
 Language: R
 Description: R is a free software environment for statistical computing and graphics.
 Author: Joe Cheng <joe@rstudio.org>
+Contributors: Konrad Rudolph <konrad.rudolph@gmail.com>
 Website: https://www.r-project.org
 Category: scientific
 */
 
 export default function(hljs) {
-  var IDENT_RE = '([a-zA-Z]|\\.[a-zA-Z.])[a-zA-Z0-9._]*';
+  // Identifiers in R cannot start with `_`, but they can start with `.` if it
+  // is not immediately followed by a digit.
+  // R also supports quoted identifiers, which are near-arbitrary sequences
+  // delimited by backticks (`…`), which may contain escape sequences. See
+  // `test/markup/r/names.txt` for examples.
+  // FIXME: Support Unicode identifiers.
+  const IDENT_RE = /(?:(?:[a-zA-Z]|\.[._a-zA-Z])[._a-zA-Z0-9]*)|\.(?!\d)|`(?:[^`]|\\`)+`/;
 
   return {
     name: 'R',
     contains: [
-      hljs.HASH_COMMENT_MODE,
```

Comment:
Typically comments always lead (to eat strings inside comments, etc), is there a reason you changed the sequencing?

Reply:
Oh, right, comments can come first (though Roxygen comments need to come even before, otherwise they won’t be recognised). It’s also important that strings come before identifiers/keywords, since raw strings start with `r`, which would otherwise be parsed as an identifier.

And keywords, in turn, need to be *inside* the main `contains` rather than outside (as other languages seem to handle this). I tried it otherwise and it led to lots of incorrect parses for number literals that weren’t fixed by liberal application of word boundaries. There’s probably a simple fix but I spent some time trying to find it without success.

Reply:
> Oh, right, comments can come first (though Roxygen comments need to come even before, otherwise they won’t be recognised). It’s also important that strings come before identifiers/keywords, since raw strings start with r, which would otherwise be parsed as an identifier.

Sounds right.

> And keywords, in turn, need to be inside the main contains rather than outside (as other languages seem to handle this). I tried it otherwise and it led to lots of incorrect parses for number literals that weren’t fixed by liberal application of word boundaries. There’s probably a simple fix but I spent some time trying to find it without success.

Ok, this is a bit strange and at least deserves a comment in the source if we're leaving it this way... can't believe i didn't notice this :-) 

>  I tried it otherwise and it led to lots of incorrect parses for number literals that weren’t fixed by liberal application of word boundaries. 

Does defining all your numbers with word boundaries not help?  Could you provide a small examples of a few failing edge cases?

Reply:
I actually managed to fix it now by restoring the mode for backtick-quoted identifiers and adding custom word boundaries to the beginning of all number modes (`(?<![a-zA-Z0-9._])`).